From 0277e3b4f8975e041f6e07a09859de9ae31db8b7 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Petr=20P=C3=ADsa=C5=99?= <ppisar@redhat.com>
Date: Wed, 4 Feb 2015 15:25:49 +0100
Subject: [PATCH] [Digest::SHA] Check for ISA when invoking methods
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

After:

use Digest::SHA;
my $d=Digest::SHA->add(qq(a));

calling $d->hashsize() and other methods resulted in crash.

This is relevant patch from Digest-SHA-5.87.

Perl RT#121421
<https://bugzilla.redhat.com/show_bug.cgi?id=1189041>

Ported to perl-5.10.1.

Signed-off-by: Petr Písař <ppisar@redhat.com>
---
 ext/Digest-SHA/SHA.xs | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/ext/Digest-SHA/SHA.xs b/ext/Digest-SHA/SHA.xs
index 7c394e9..980a5e6 100644
--- a/ext/Digest-SHA/SHA.xs
+++ b/ext/Digest-SHA/SHA.xs
@@ -152,6 +152,8 @@ PREINIT:
 	SHA *state;
 	int result;
 PPCODE:
+	if (!sv_isa(self, "Digest::SHA"))
+		XSRETURN_UNDEF;
 	state = INT2PTR(SHA *, SvIV(SvRV(SvRV(self))));
 	result = shadsize(state) << 3;
 	if (ix == 1 && result == 160)
@@ -168,6 +170,8 @@ PREINIT:
 	STRLEN len;
 	SHA *state;
 PPCODE:
+	if (!sv_isa(self, "Digest::SHA"))
+		XSRETURN_UNDEF;
 	state = INT2PTR(SHA *, SvIV(SvRV(SvRV(self))));
 	for (i = 1; i < items; i++) {
 		data = (unsigned char *) (SvPV(ST(i), len));
@@ -187,6 +191,8 @@ PREINIT:
 	SHA *state;
 	char *result;
 PPCODE:
+	if (!sv_isa(self, "Digest::SHA"))
+		XSRETURN_UNDEF;
 	state = INT2PTR(SHA *, SvIV(SvRV(SvRV(self))));
 	shafinish(state);
 	len = 0;
-- 
1.9.3

